test/e2e/framework/kubectl/kubectl_util.go

```diff
@@ -0,0 +1,438 @@
+/*
+Copyright 2014 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubectl
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/url"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/onsi/ginkgo/config"
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	utilversion "k8s.io/apimachinery/pkg/util/version"
+	clientset "k8s.io/client-go/kubernetes"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2elog "k8s.io/kubernetes/test/e2e/framework/log"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	uexec "k8s.io/utils/exec"
+)
+
+const defaultHost = "http://127.0.0.1:8080"
+
+var (
+	// For parsing Kubectl version for version-skewed testing.
+	gitVersionRegexp = regexp.MustCompile("GitVersion:\"(v.+?)\"")
+	// TODO: break out common logic with framework/test_context.go, to separate sub-package within framework
+	testContext testContextType
+)
+
+// TODO: break out common logic with framework/test_context.go, to separate sub-package within framework
+type testContextType struct {
+	KubeConfig  string
+	KubeContext string
+	CertDir     string
+	Host        string
+	CloudConfig cloudConfig
+	KubectlPath string
+	// Provider identifies the infrastructure provider (gce, gke, aws)
+	Provider             string
+	AllowedNotReadyNodes int
+}
+
+// TODO: break out common logic with framework/test_context.go, to separate sub-package within framework
+type cloudConfig struct {
+	APIEndpoint       string
+	ProjectID         string
+	Zone              string // for multizone tests, arbitrarily chosen zone
+	Region            string
+	MultiZone         bool
+	MultiMaster       bool
+	Cluster           string
+	MasterName        string
+	NodeInstanceGroup string // comma-delimited list of groups' names
+	NumNodes          int
+	ClusterIPRange    string
+	ClusterTag        string
+	Network           string
+	ConfigFile        string // for azure and openstack
+	NodeTag           string
+	MasterTag         string
+
+	Provider framework.ProviderInterface
```

Comment:
How do handle this logic @alejandrox1 ? 

I put `framework.ProviderInterface` just to get this to build, but once I put `framework/kubectl` as a required import in `framework/util.go` it will all fail. The issue I am running into is that `TestContext` is required by `KubectlCmd`. I cant just import `framework` because that will cause circular dependencies as described above.

I've tried copying logic from `TestContext` like with `ExpectNoError`, unexporting it and using it locally in this file.

It seems to work ok except for this `ProviderInterface`, which again, belongs to `framework` because of `framework/provider.go`. Does it make sense to try and unexport the `ProviderInterface` and add it to this file?

It seems I keep going this rabbit hole of unexporting stuff just to use 1 or 2 functions "/. What's a good "Go" way to do this? Thank you for all of your help btw!

Reply:
@alejandrox1 TLDR; I need to find a way to avoid importing `framework` in `framework/kubectl/kubectl.go` because I have to import `framework/kubectl` inside of `framework/kubectl/util.go`, for various reasons (a lot of functions use it).

Reply:
@MorrisLaw did you solve this? This bit is outdated and I can't find the original patch you proposed :sweat_smile: 
I don't think we discussed this bit in slack but please feel free to ping me here or on slack if the `ProviderInterface` is something you still need and are having issues with so we can figure it out :slightly_smiling_face: